### PR TITLE
Add exomol support

### LINKED ISSUE #319: ExoMol Support
### 🎯 Describe what you'd like

ExoMol support. Asked for in #34 

### Implementation (see https://github.com/radis/radis/pull/320)

```python
from radis import SpectrumFactory

sf = SpectrumFactory(
        4310,
        4320,
        molecule="H2O",
        isotope="1",
        verbose=3,
    )
sf.fetch_databank("exomol")
s = sf.eq_spectrum(500)
s.plot()
```

image

### 💡 Possible implementations

Borrowing and adapting some codes from the [ExoJax](https://github.com/HajimeKawahara/exojax) project from @HajimeKawahara (which you should also have a look at !) Hopefully we can end up maintaining a unified code base for local database formats (currently : ExoJax uses Feather format; which is extremelly efficient but cannot be masked/filtered as easily as HDF5)

---
<sub>*Tip : you can vote for [New Features](https://github.com/radis/radis/issues/79) on https://feathub.com/radis/radis*</sub>

### PR DESCRIPTION
**Pull request by erwanp**
### Description

First spectrum computed from ExoMol; borrowing some ExoJax codes from @HajimeKawahara  

```python
from radis import SpectrumFactory

sf = SpectrumFactory(
        4310,
        4320,
        molecule="H2O",
        isotope="1",
        verbose=3,
    )
sf.fetch_databank("exomol")   # recommended ExoMol database; or choose exomol_database=... to choose another
s = sf.eq_spectrum(Tgas=500)
s.plot()
```

image

Still WIP,

- [x] read local ExoMol files
- [ ] convert them to RADIS's HDF5 format (Pytables #203 switching to Vaex #306 #295 )
- [x] load & convert to Radis's default Pandas DataFrame
- [x] compute missing values (Linestrengths, energies)
- [x] compute a Spectrum from ExoMol files
- [ ] add isotope abundances : [here](https://github.com/erwanp/radis/blob/add/exomol_support/radis/io/exomol.py#L391)
- [ ] add isotope number: [here](https://github.com/erwanp/radis/blob/add/exomol_support/radis/io/exomol.py#L590) and [there](https://github.com/erwanp/radis/blob/add/exomol_support/radis/io/exomol.py#L299)
- [ ] read molecular parameters from ExoMol (and update MolParams, currently = HITRAN data only. Requires #321 )
- [ ] multiple isotope support
- [ ] superline support

Interface : 
- [x] auto-download ExoMol files
- [x] auto-find ExoMol full names, and available databases
- [x] update SpectrumFactory fetch_databank
- [ ] update SpectrumFactory load_databank functions
- [ ] add explicit message to fill [KNOWN_EXOMOL_ISOTOPES_NAMES](https://github.com/radis/radis/blob/92fceb691e3695da25f8c78f3df28f487b5325d8/radis/db/classes.py#L280) if molecule not supported but exists in ExoMol
- [x] auto-find all ExoMol molecules by parsing [this page](https://exomol.com/data/molecules/).

Tests / Docs : 

- [x] add unitary tests
- [ ] add comparison with HITRAN/HITEMP
- [x] add a Gallery Example

Fixes #319

### CODE REVIEW

File: radis/db/molparam.py [hunk 1]
```diff
@@ -192,7 +192,7 @@ def get(self, M, I, key):
             molecule id
         I: int
             molecule isotope #
-        key: ``'abundance'``, ``'mol_mass'``
+        key: ``'abundance'``, ``'mol_mass'``, ``'isotope_name_exomol'``,
```
Comment (erwanp) [inline radis/db/molparam.py:195]:
```suggestion
        key: ``'abundance'``, ``'mol_mass'``, ``'isotope_name'``, ``'isotope_name_exomol'``, 
```

I also added this one

**Comment by erwanp** [comment]
New improvements in the ExoMol implementation, we are now able to compute non-HITRAN molecules (with some hardcoded parts, still). Interface is also improved, and as easy as 'hitran' / 'hitemp' . See here a SiO2 spectrum from the EBJT Exomol : 

```python
from radis import calc_spectrum
s = calc_spectrum(1080, 1320,         # cm-1
                  molecule='SiO',
                  isotope='1',
                  pressure=1.01325,   # bar
                  Tgas=1000,           # K
                  mole_fraction=0.1,
                  path_length=1,      # cm
                  databank='exomol',  # or use ('exomol', "EBJT") for a specific database ("EBJT")
                  )
s.apply_slit(1, 'cm-1')       # simulate an experimental slit
s.plot('radiance')
```
image

File: radis/lbl/loader.py [hunk 2]
```diff
@@ -810,9 +823,9 @@ def fetch_databank(
             dbformat = "exomol-radisdb"  # downloaded in RADIS local databases ~/.radisdb  # Note @EP : still WIP.
         if exomol_database != None:
             assert source == "exomol"
-        if parfuncfmt == "exomol" and source != "exomol":
+        if [parfuncfmt, source].count("exomol") == 1:
             raise NotImplementedError(
-                "ExoMol partition functions must be used with ExoMol database"
+                "ExoMol partition functions must be used with ExoMol database - and vice-versa"
```
Comment (erwanp) [inline radis/lbl/loader.py:828]: Actually it worked the other way around, but you're right, let's be consistent and enforce consistency!

File: radis/lbl/base.py [hunk 3]
```diff
@@ -2017,49 +2017,44 @@ def _calc_Q(molecule, iso, state):
 
         if "id" in df1.columns:
```
Comment (minouHub) [inline radis/lbl/base.py:2018]: It is not possible to calculate HITRAN simulation because of this change

**Review by minouHub — CHANGES_REQUESTED** [review]

File: radis/lbl/base.py (hunk 3 above)
Comment (minouHub) [reply radis/lbl/base.py:2018]: sorry but I didn't really get what you wanted to do on this line so I hardly try to mess up

Comment (erwanp) [reply radis/lbl/base.py:2018]: Do you get AttributeError? 
If so, we could switch it to a self.warn(..., "PerformanceWarning") 

Id is the molecule number. For performance reasons it shouldn't be stored as a column, this takes unnecessary Memory space.  It's related to @gagan-aryan project. But until the column is removed and replaced with a simple attribute of the dataframe, having a warning should still allow you to compute HITRAn

Comment (minouHub) [reply radis/lbl/base.py:2018]: thank you for doing the correction :)

**Comment by minouHub** [comment]
**First, good news:** nice comparison of H2CO in ExoMol and HITRAN. I will add it somewhere in the week.
image

**Now, help wanted @erwanp @gagan-aryan :**
I need to import other quantum numbers in the df0. To give you an example, the broadening of H2CO is function of J and K. This import of quantum numbers takes a huge memory (my 16 GB of RAM are full during the process). The memory overload happens here: https://github.com/radis/radis/blob/ed74864355c0ad4be21f34f06ec4cf854f6f2f1a/radis/io/exomolapi.py#L241

I plan to improve that. First:

- [ ] The number of lines should be diminished before this step. 

I really believe that I handle these data not efficiently. Otherwise, other ideas would be:

- [ ] add an argument in the fetch_database to let the user import the quantum number of his choice

- [ ] [your ideas]

**Comment by erwanp** [comment]
This first implementation ExoMol implementation was borrowed from Exojax and used the Feather format. We'll switch to Vaex's HDF5 where your `n0 = 4 + dic_def['Landé'] + dic_def['lifetime']` definition can become a **virtual column**; i.e.; you'll use the same syntax to define the column, but the formula is evaluated only for the lines in your spectral range : no more memory problems.

**Review by gagan-aryan — COMMENTED** [review]
I have check all of the files thoroughly and lgtm. If I come across any issue while working on these files, I will push the code.

**Comment by minouHub** [comment]
Impossible to convert the trans.bz2 file of CH4 to a feather
**Code:**

```python
from radis import SpectrumFactory
molecule = "CH4"

conditions = {
    "mole_fraction": 0.01,
    "pressure": 1,
    "Tgas": 700,
}
# wmin, wmax = 2050, 2060
wmin, wmax = 2777, 2779

sf = SpectrumFactory(
        wmin=wmin,
        wmax=wmax,
        molecule=molecule,
        isotope="1",
        verbose=3,
        wstep = 2e-4,
    )
sf.fetch_databank("exomol", parfuncfmt = "exomol")
```

The file is just 1.7 GB! 

```python
MemoryError: Unable to allocate 2.85 GiB for an array with shape (2, 191003050) and data type float64
```

**Comment by erwanp** [comment]
@gagan-aryan is currently working on writing a vaex-HDF5 builder for HITEMP #306 . We'll use the same to replace the Feather building step.

**Comment by erwanp** [comment]
About to merge all current implementation. There are many changes already and we don't want to diverge too much from the rest of the code.

Some things aren't final (see checkbox above) but will be finished in a later PR.

**Pull request merged**